homeassistant/components/nws/sensor.py

```diff
@@ -0,0 +1,134 @@
+
+ALERT_VALID_TIME = timedelta(minutes=10)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the NWS sensors."""
+
+    hass_data = hass.data[DOMAIN][config_entry.entry_id]
+    nws = hass_data[NWS_DATA]
+    latitude = config_entry.data[CONF_LATITUDE]
+    longitude = config_entry.data[CONF_LONGITUDE]
+    coordinator = hass_data[COORDINATOR_ALERTS]
+
+    await coordinator.async_refresh()
+    if not nws.all_zones:
+        raise PlatformNotReady
+
+    async_add_entities(
+        [
+            NWSAlertSensor(
+                f"{base_unique_id(latitude, longitude)}_alerts", coordinator, nws,
+            ),
+        ],
+        False,
+    )
+
+
+class NWSAlertSensor(Entity):
+    """Representation of a sensor entity for NWS alert values."""
+
+    def __init__(self, unique_id, coordinator, nws):
+        """Initialize the sensor."""
+        self._unique_id = unique_id
+        self._coordinator = coordinator
+        self._nws = nws
+
+        self._zones = sorted(nws.all_zones)
+        self._name = f"{' '.join(self._zones)} Alerts"
+
+        self._alerts = None
+
+    @property
+    def unique_id(self):
+        """Sensor Unique id."""
+        return self._unique_id
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return "mdi:weather-cloudy-alert"
+
+    @property
+    def state(self):
+        """Return entity state as number of alerts."""
+        return len(self._alerts)
+
+    @property
+    def should_poll(self):
+        """No need to poll. Coordinator notifies entity of updates."""
+        return False
+
+    @property
+    def available(self):
+        """Return if entity is available."""
+        if self._coordinator.last_update_success_time:
+            last_success_time = (
+                utcnow() - self._coordinator.last_update_success_time < ALERT_VALID_TIME
+            )
+        else:
+            last_success_time = False
+        return self._coordinator.last_update_success or last_success_time
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit_of_measurement."""
+        return "alerts"
+
+    @property
+    def device_state_attributes(self):
+        """Return the sensor attributes."""
+        return {
+            ATTR_ALERTS: self._alerts,
+            ATTR_ATTRIBUTION: ATTRIBUTION,
+            ATTR_ZONES: self._zones,
+        }
+
+    async def async_update(self):
+        """Update the entity.
+
+        Only used by the generic entity update service.
+        """
+        await self._coordinator.async_request_refresh()
+
+    @callback
+    def _update_state(self):
+        if self._coordinator.data:
+            self.hass.bus.fire(
+                EVENT_ALERTS,
+                {ATTR_ALERTS: self._coordinator.data, ATTR_ZONES: self._zones},
+            )
```

Comment:
We shouldn't double up on an event and a state change since that means we fire two events every time this changes.  We should do one or the other.

Reply:
As stated above, it is common to have one event replace another, for example, a thunderstorm warning may replace a thunderstorm watch. I want to fire a notification when a warning is given.  It is very hard to do in this situation without a separate event.  It is even harder when there are more than one alerts to make automations.

You are suggesting just to fire events, which solves this, but now the user cannot access the current alerts without setting up a complex set of automations, templates, etc.  I could fire "alert expired events" so that users could use them to clean up their information, but this situation seems worse to me than firing one more event in the base integration.

Reply:
If the worry is that this is purely duplicated information, the state has all alerts that are still current, but the event only contains the new events.

Reply:
If I have to choose for myself and how I usr it,  I would use only events.

Reply:
State changed events are roughly three as expensive as other events to fire.   They also use a little more than twice the space in the database since they store a row in the states table as well.

This is one of the reasons we have a policy against firing events from entities.

Reply:
I can refactor to fire the event in the coordinator update, but I'm getting the impression that another approach is preferred.  Would it be preferred to implement a core integration like the geo_location interegration with individual entities for each alert?  This seems to move in the opposite direction since there would be a state change event and a new entity created (I assume this is even more costly) with every alert.  On the other hand, it would be a base for other interegrations and there is precedence for this structure.

Reply:
@MartinHjelmare When you have a free moment, would you kindly weigh-in on this?

Reply:
Could we make a geo_location platform that creates an entity for each geocode in the alert and calculates the distance from the user to that geocode?

The geo_location platform should keep track of existing alerts and care for removing stale alert entities.

Reply:
The location in this type of alert is for an entire zone (usually), which is a predetermined polygonal shape, sometimes there is a specific polygon provided.  There is no center location that the geo_locatiom integration wants.  I suppose that the distance could always be zero?

Reply:
I should have mentioned that the user will always be in the polygon.

Reply:
Ok. Then I'd recommend just firing custom events and not do an entity.

Reply:
I think the geo_location platform will work if we set the location of every alert to be concident on the configured location for the alert.  This is consistent with the incoming data and the way it is used in home assistant.

Reply:
If we calculate the distance between the set coordinates and the user home coordinates, that could make sense.

Reply:
I will mark this draft in the meantime, while I convert to geo_location.  The user is almost always in the polygon, or is very close, so an MVP could be to set the distance to be zero.  I'll plan on adding "distance to polygon" in the upstream pypi location so that it the future the actual distance can be used.